Source/GSAutoLayoutEngine.m

```diff
@@ -22,39 +22,432 @@
+    (NSLayoutConstraint *)constraint;
+
+- (void) addConstraintAgainstViewConstraintsArray:
+    (NSLayoutConstraint *)constraint;
+
+- (void) removeConstraintAgainstViewConstraintsArray:
+    (NSLayoutConstraint *)constraint;
+
+- (void) addSupportingInternalConstraintsToView:(NSView *)view
+                                  forAttribute:(NSLayoutAttribute)attribute
+                                    constraint:(GSCSConstraint *)constraint;
+
+- (void) removeInternalConstraintsForView:(NSView *)view;
+
+- (void) addSolverConstraint:(GSCSConstraint *)constraint;
+
+- (void) removeSolverConstraint:(GSCSConstraint *)constraint;
+
+- (NSArray *) constraintsForView:(NSView *)view;
+
+- (NSNumber *) indexForView:(NSView *)view;
+
+- (void) notifyViewsOfAlignmentRectChange:(NSArray *)viewsWithChanges;
+
+- (BOOL) solverCanSolveAlignmentRectForView:(NSView *)view solution:(GSCSSolution *)solution;
+
+- (void) recordAlignmentRect:(NSRect)alignmentRect forViewIndex:(NSNumber *)viewIndex;
+
+- (NSRect) solverAlignmentRectForView:(NSView *)view solution:(GSCSSolution *)solution;
+
+- (NSRect) currentAlignmentRectForViewAtIndex:(NSNumber *)viewIndex;
+
+- (BOOL) updateViewAligmentRect:(GSCSSolution *)solution view:(NSView *)view;
+
+@end
 
 @implementation GSAutoLayoutEngine
 
-- (void) addConstraint: (NSLayoutConstraint *)constraint
+- (instancetype) initWithSolver:(GSCassowarySolver *)cassowarySolver;
+{
+  self = [super init];
+  if (self != nil)
+    {
+      _viewCounter = 0;
+      _solver = cassowarySolver;
+      RETAIN (_solver);
+
+      _variablesByKey = [NSMapTable strongToStrongObjectsMapTable];
+      RETAIN (_variablesByKey);
+
+      _constraintsByAutoLayoutConstaintHash =
+          [NSMapTable strongToStrongObjectsMapTable];
+      RETAIN (_constraintsByAutoLayoutConstaintHash);
+
+      _layoutConstraintsBySolverConstraint =
+          [NSMapTable strongToStrongObjectsMapTable];
+      RETAIN (_layoutConstraintsBySolverConstraint);
+
+      _solverConstraints = [NSMutableArray array];
+      RETAIN (_solverConstraints);
+
+      _trackedViews = [NSMutableArray array];
+      RETAIN (_trackedViews);
+
+      _supportingConstraintsByConstraint =
+          [NSMapTable strongToStrongObjectsMapTable];
+      RETAIN (_supportingConstraintsByConstraint);
+
+      _viewAlignmentRectByViewIndex = [NSMutableDictionary dictionary];
+      RETAIN (_viewAlignmentRectByViewIndex);
+
+      _viewIndexByViewHash = [NSMutableDictionary dictionary];
+      RETAIN (_viewIndexByViewHash);
+
+      _constraintsByViewIndex = [NSMutableDictionary dictionary];
+      RETAIN (_viewIndexByViewHash);
+
+      _internalConstraintsByViewIndex =
+          [NSMapTable strongToStrongObjectsMapTable];
+      RETAIN (_internalConstraintsByViewIndex);
+    }
+  return self;
+}
+
+- (instancetype) init
 {
-  // FIXME Add constraint to solver
+  GSCassowarySolver *solver = AUTORELEASE ([[GSCassowarySolver alloc] init]);
+  return [self initWithSolver: solver];
```

Comment:
To avoid the usage of the autorelease pool I would write this somewhat like
```
GSCassowarySolver *solver = [[GSCassowarySolver alloc] init];
self =  [self initWithSolver: solver];
RELEASE(solver);
return self;
```